### PR TITLE
Adds a PR guidelines section to contributing.md

### PR DESCRIPTION
**Pull request by dakinggg**
# What does this PR do?
Adds a section of `CONTRIBUTING.md` with some basic PR guidelines that we would like to be more rigorous about enforcing.

# What issue(s) does this change relate to?
Closes [CO-1820](https://mosaicml.atlassian.net/browse/CO-1820)

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [x] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.

[CO-1820]: https://mosaicml.atlassian.net/browse/CO-1820?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by mvpatel2000 — APPROVED** [review]
lgtm. I will be better at this as I am very guilty of failing certain points :(

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -55,6 +55,14 @@ git checkout -b cool-new-feature
 
 4\. When you are ready, submit a pull request into the composer repository! If merged, we'll reach out to send you some free swag :)
 
+## Pull request guidelines
+
+We have some rough guidelines that will make your PR easier to review and more likely to get smoothly merged.
+* Self-contained title and description. Please include a descriptive title and description. These should allow someone to understand the contents of the PR without looking at the code.
```
Comment (bandish-shah) [inline CONTRIBUTING.md:61]: The title should be descriptive yet concise.  Ideally the title should answer the question, "what does this PR change or do?".  There should be enough information that users can understand at a glance and if they're interested can click to read additional details.

For the PR description I'd like to emphasize the bit that user's (technical and possibly semi-technical) can understand what a given PR does _without_ looking at the code.  In plain English, the description should highlight what the PR changes and what the impact to user's will be (if any).

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -55,6 +55,14 @@ git checkout -b cool-new-feature
...
+* If the PR affects output that is displayed to a user of Composer (e.g. console logging or experiment tracker reporting), please include screenshots showing what the new output looks like. UX is important!
+* Include tests. If you are fixing a bug, please add a test that would've caught the bug. If you are adding a new feature, please add unit tests that test the various components of the feature, and also a test that tests the full functionality of the feature.
```
Comment (bandish-shah) [inline CONTRIBUTING.md:63]: Should we include something about manual testing as well?  For example, we don't run the full test suite in PR because we want PR CI tests to run quickly.  Ideally it would be good if developers ran other tests that aren't run by the PR CI (e.g., things we run in the `daily`).  I don't think this is a requirement as there's no easy way to enforce it other than just running the tests, but it is good practice.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -55,6 +55,14 @@ git checkout -b cool-new-feature
...
+* `pre-commit` should help you handle formatting and type checking, but please do make sure you have it installed as described [above](#prerequisites).
+
```
Comment (bandish-shah) [inline CONTRIBUTING.md:65]: I'm thinking about enabling this soon:
https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes

Just mentioning here because we can start using labels as well and it will greatly simplify creating release notes.

**Review by eracah — APPROVED** [review]
LGTM

File: CONTRIBUTING.md (hunk 3 above)
Comment (dskhudia) [reply CONTRIBUTING.md:65]: Also we should mention what is must and nice-to-have out of these. Basically we don't want to give a big list of things to do to external contributors.

**Review by bcui19 — APPROVED** [review]
LGTM thank you for helping make our review processes better

**Review by karan6181 — APPROVED** [review]

**Review by bandish-shah — APPROVED** [review]
Signing my soul over to good PR descriptions and details!

**Review by knighton — APPROVED** [review]

File: CONTRIBUTING.md (hunk 3 above)
Comment (dakinggg) [reply CONTRIBUTING.md:65]: I consider all of these things essentially required (except running daily tests), but I will add some language up front saying that any uncertainty/difficulty with any of these should not deter from opening a PR and we are happy to help
